### DISCUSSION TITLE
Localized Text Field Component

### DISCUSSION THREAD

**Discussion by manuelebagnolini** [payloadcms/payload#1308] [Feature Requests & Ideas]
Hi,

we are a web agency with more than 20 years of experience and we are deciding to replace our custom-made CMS with a modern headless CMS, and we’ll really like to use Payload CMS.

The only feature we are missing is the possibility to have a localized field component that permits to:

- Instantly understand which fields are localizable and which are not.
- Instantly understand which localizations are provided for the specific field (without having to switch locale for the full document).
- Cycle between field localizations to easily check the current values.
- Facilitate quick data entry even in the case of a large amount of values at once.
- Easily add a localized value for a language not yet provided.

This is an example of how our custom CMS actually works:

1. The default language (English) value is shown in the field header and in the input field of the English tab. Just by looking at this, the user knows that this is a localized text field and that it’s currently localized in English and Italian.
![cms_loc_text_01](https://user-images.githubusercontent.com/45173102/198567779-b7519a55-2753-41e2-87a7-2f4d564651b7.jpg)

1. By selecting the Italian tab the user can see and edit the Italian localization.
![cms_loc_text_02](https://user-images.githubusercontent.com/45173102/198567852-af6f4b6d-3004-4ec0-aeec-fb703183292d.jpg)

1. By selecting the feature “Add text in a language” the user can select one of the available languages not yet used for the field, in this example the French tab will be added and the user could provide the French localized value.
![cms_loc_text_03](https://user-images.githubusercontent.com/45173102/198567913-a23d5ffb-0a2a-4ffe-87b2-e56294617705.jpg)

We are used to build complex sites with a lot of different languages (usually more than 10) and this feature is indispensable for us for choosing our new CMS.

Our customers find this system very intuitive, and we think it would be a great addition to Payload CMS. 

Obviously, we don’t expect the component to be exactly like the one of our custom-made CMS, the pictures are just an example to explain the feature, the new component should be integrated with the style and UX of the Payload CMS Admin UI.

You could be interested to include this feature in Payload CMS?

In case we would be available to actively collaborate in the development.

Looking at the code, we should be able to create our own custom UI component but the problem is that the field value of the localized field is flattened to the current locale and we can’t load and save the JSON data containing all the languages at once.

Our basic idea is to create a new version of the `useField` hook (i.e., `useLocalizedField`) that works with not flattened localized fields like in the `beforeRead` collection hook. We understand that is a major change in the Payload CMS core regarding the data access of the collections.

**Comment by jmikrut** [comment]
Hey @manuelebagnolini — thank you for going into such detail here. We really appreciate it!

We're actually going to be working on improving our localization experience over the next few months, and I think we can make a lot of this easier for you and your team. Many of the ideas that you're referencing could end up being possible for sure.

> we can’t load and save the JSON data containing all the languages at once.

Actually, you can! Use `locale=*` as a query param in the REST API and you can get data for all locales. 👍  That might help you get started on your custom UI fields at least. Also, I should point out that you can probably get some fancy stuff done with `Description` components as well, on each field, so as to try and indicate which field has which translation, etc.

As for the concrete items that we plan to take on and build, I will update this thread once we have had time to nail them all down. But this is good stuff.

**Reply by manuelebagnolini** [reply]
Thank you @jmikrut.
In the meantime we'll look into your suggestions.👍

**Comment by manuelebagnolini** [comment]
I share an update on the development of the feature.

![PayloadLocalizationUIField](https://user-images.githubusercontent.com/45173102/202403083-d736e765-2dba-424f-a6b2-10cbe6b164e0.gif)

In details we added:
- A sticky component at the top of document with the locale selector to easily switch locales.
- A "Change localization" button to instantly understand which field are localized and access the localization feature
- A modal that allows you to edit the existing localizations and insert new ones. The localization in the reference language (current document locale) is always visible to facilitate the translation. You can save all changed localizations at once as a drafts or publish them directly.

We are currently having problems saving so we opened these issues: #1415, #1419 

Instead of calling an API for each locale it would be more efficient to call a single API to massively update all languages. Could it be added to this discussion: #1382?

**Reply by b-barry** [reply]
Are you planning to make it open source?

**Reply by manuelebagnolini** [reply]
> Are you planning to make it open source?

Yes, it's part of an open source project that we plan to release this year.

**Comment by ssyberg** [comment]
@manuelebagnolini if I'm understanding correctly you're developing this feature outside of the mainline payload project correct? Do you have any indication if this will be brought in to the core?

**Reply by manuelebagnolini** [reply]
We are developing this feature as a plugin so it can be installed and configured alongside the other payload plugins.
I have no indication of future implementations in the payload core of a similar feature.

**Reply by ssyberg** [reply]
Cool thank you! Any idea on timeline? We're just deciding if we should implement our own...

**Reply by manuelebagnolini** [reply]
Our goal is to publicly release our main project, that this plugin is a part of, this year.
I don't have a target date for the release yet but I will update this discussion when I have precise information.

**Reply by ssyberg** [reply]
@manuelebagnolini how is this coming along?

**Reply by manuelebagnolini** [reply]
We are still developing the main project but you can try an **alpha version** of this plugin.

The source code is inside the main project monorepo which is still private at the moment.

Given its alpha nature it is necessary to make some tweaks to the configuration, in the next versions we will improve this aspect.
In particular you need to add a webpack specific configuration and make sure that globals and collections are not _null_/_undefined_ but at least empty list.

Install:

```bash
npm i @websolutespa/payload-plugin-localization
```

```bash
npm i path-browserify
```

Add to _payload.config.ts_:

```ts
import { buildConfig } from 'payload/config';
import { localization } from '@websolutespa/payload-plugin-localization';
import '@websolutespa/payload-plugin-localization/dist/index.css';

export default buildConfig({
  collections: [],
  globals: [],
  admin: {
    webpack: (config) => {
      config.resolve = {
        ..config.resolve,
        fallback: {
          "fs": false,
          "path": require.resolve("path-browserify"),
          "crypto": false,
        }
      }
    return config;
  },
  plugins: [
    localization(),
  ],
}
```

You can find a description of how it works [here](https://www.npmjs.com/package/@websolutespa/payload-plugin-localization).

If you have any feedback, suggestions or criticisms we would love to hear from you.

**Reply by ssyberg** [reply]
@manuelebagnolini haven't had a chance to dive in too deep, but I can report it installed seamlessly into my project and the UI is very cool! One thing we are trying to figure out is how better to highlight non-localized fields to editors so they don't accidentally make a change they don't realize will affect all locales... I also want a "translation mode" where non-localized fields are disabled or hidden?

**Reply by manuelebagnolini** [reply]
@ssyberg thanks for the feedback, I shared it with the dev team and we'll talk about it in future brainstormings.

**Reply by ssyberg** [reply]
> A sticky component at the top of document with the locale selector to easily switch locales (for all collections even if they do not contain localized fields).

@manuelebagnolini a little more feedback, our client is finding the stickiness of this a bit confusing on sections of the CMS that have no localization and I've been asked to unsintall the plugin until we can address it - very promising, looking forward to seeing how this evolves!

**Reply by manuelebagnolini** [reply]
Hi @ssyberg thanks again for your feedback!

We have just released a new version (0.0.2) that solves the problem you reported and also simplifies the configuration of the plugin.
It is no longer necessary to manually specify the webpack configuration (no need to install manually _path-browserify_ too) and the _collections_ and _globals_ parameters can now be omitted if not needed.

Install:

```bash
npm i @websolutespa/payload-plugin-localization
```

Add to _payload.config.ts_:

```ts
import { buildConfig } from 'payload/config';
import { localization } from '@websolutespa/payload-plugin-localization';
import '@websolutespa/payload-plugin-localization/dist/index.css';

export default buildConfig({
  plugins: [
    localization(),
  ],
}
```

**Reply by ssyberg** [reply]
Thanks for the update! I actually tried upgrading to 0.0.2 but now the entire language selector is gone.

**Reply by manuelebagnolini** [reply]
Hi @ssyberg sorry to hear that.

I just tried both updating an existing project and installing it on a new project and it works correctly in both cases showing the language selector only for collections that have at least one field with `localized: true`.

Can you share us an example or sample code?

**Reply by piccinnigius** [reply]
Hi @manuelebagnolini ,  i really appreciate this plugin and it fit perfectly what we need, thanks! I'll let you know any feedback

**Reply by simonereda** [reply]
Hi @manuelebagnolini , thank you for your plugin! It's very helpful! I would like to try adding a button that allows me to translate the text using ChatGPT, but I can't open your code repository. Will you share the repo?

**Reply by ssyberg** [reply]
@manuelebagnolini is it possible it doesn't play nicely with tabs? I just upgraded to 0.0.3 now and this is what I'm seeing:
<img width="451" alt="image" src="https://github.com/payloadcms/payload/assets/1661046/968cc4ba-6f17-4da6-bd80-6be9ea881532">

Code:
```
const Sandbox = {
    slug: 'sandbox',
    fields: [
        {
            name: 'test',
            type: 'text',
            localized: true,
        },
    ],
};

export default Sandbox;
```

**Reply by ssyberg** [reply]
Oh looks like this import is still required! 
`import '@websolutespa/payload-plugin-localization/dist/index.css';`
Added that back in and its working now

**Reply by ssyberg** [reply]
@manuelebagnolini ok so I've narrowed down the issue I was having above and it looks like this is conflicting with another plugin we use that implements collection inheritance.

This works fine:
```
const Sandbox = {
    slug: 'sandbox',
    fields: [
        {
            name: 'test',
            type: 'text',
            localized: true,
        },
    ],
};

export default Sandbox;
```
while this removes the main language selector above the tabs:

```
const Sandbox = {
    slug: 'sandbox',
    custom: { extendsBase: ['base'] },
    fields: [
        {
            name: 'test',
            type: 'text',
            localized: true,
        },
    ],
};

export default Sandbox;
```
The only different is the inheritance plugin is invoked. That plugin runs _before_ your plugin so you'd think they'd be safe to use together. That "base" adds a few fields (mostly to the sidebar) and tabs to the collection, but hard to say what else could be happening that messes this up without seeing the plugin code... anyways I guess not necessarily a bug some change between 0.0.1 and 0.0.2 introduced this.

**Reply by ssyberg** [reply]
A bit of more info, when I put your plugin before the inheritance plugin, it shows up but not in the right place haha - I can do a quick fix here but I'm guessing there's something about the tab and sidebar field configuration that we're doing that your plugin doesn't like because I can see the config doesn't include your `uiLanguages` field if I use it like I am doing above.
![image](https://github.com/payloadcms/payload/assets/1661046/63ad195d-2454-4cde-a5fa-8c69a4f8fe9e)

**Reply by ssyberg** [reply]
Alright I was able to dive into the `dist` and figure this out, there's something about how you're determining localized field existence that isn't working for our particular configs (it appears to have something to do with our tabs but I'm not totally sure). I do however have a quick fix I believe. Instead of attempting to traverse the entire field tree in all of it's potentially structures, this is what I've done in other plugins to look for the existence of something:

`JSON.stringify(config.fields).includes('"localized":true')`

There's probably some weird edge case where this is not strictly safe (you have that exact string in some `description` field but you don't actually have any localized fields?) but it's fast and it would certainly solve my current bug :)

**Reply by manuelebagnolini** [reply]
Hi @ssyberg, thank you for the update!

For the direct css import in the config file:
```ts
import '@websolutespa/payload-plugin-localization/dist/index.css';
```
As you said it's still needed but the team is planning to omit it in the next releases.

For the fix that you made:
```ts
JSON.stringify(config.fields).includes('"localized":true')
```
Could you tell us which plugin it's in conflict with and its configuration so that we can try it too?

**Reply by ssyberg** [reply]
Unfortunately it's also a custom plugin in a monorepo. I will see if I can get you a failing collection config to work from though.

**Reply by ssyberg** [reply]
@manuelebagnolini here's an example config, can't explain why but I've verified that `hasLocalizedFields` is false here.

```
config.fields [
  {
    "name": "editSyndication",
    "type": "ui",
    "admin": {
      "position": "sidebar",
      "components": {}
    }
  },
  {
    "type": "row",
    "label": " ",
    "fields": [
      {
        "name": "slug",
        "type": "text",
        "required": true,
        "hooks": {
          "beforeValidate": [
            null
          ]
        },
        "admin": {
          "description": "If left blank will autogenerate from the title."
        }
      },
      {
        "name": "path",
        "type": "text",
        "unique": true,
        "admin": {
          "readOnly": true,
          "position": "sidebar"
        },
        "hooks": {
          "beforeChange": [
            null,
            null
          ]
        }
      }
    ],
    "admin": {
      "position": "sidebar"
    }
  },
  {
    "name": "translations",
    "type": "select",
    "options": [
      {
        "label": "En",
        "value": "English"
      },
      {
        "label": "Ar",
        "value": "Arabic"
      },
      {
        "label": "Cr",
        "value": "Cyrillic"
      },
      {
        "label": "Hi",
        "value": "Hindi"
      },
      {
        "label": "Ch",
        "value": "Chinese"
      },
      {
        "label": "Sp",
        "value": "Spanish"
      },
      {
        "label": "Ge",
        "value": "German"
      },
      {
        "label": "Fr",
        "value": "French"
      },
      {
        "label": "Th",
        "value": "Thai"
      }
    ],
    "hasMany": true,
    "admin": {
      "readOnly": true,
      "position": "sidebar"
    }
  },
  {
    "name": "index",
    "type": "relationship",
    "label": "Syndications",
    "relationTo": "indexes",
    "hasMany": false,
    "admin": {
      "hidden": true
    }
  },
  {
    "localized": true,
    "admin": {
      "components": {}
    },
    "name": "title",
    "type": "text",
    "required": true,
    "unique": true
  },
  {
    "name": "_collection",
    "type": "text",
    "admin": {
      "hidden": true
    }
  },
  {
    "localized": true,
    "admin": {
      "components": {}
    },
    "name": "test",
    "type": "text"
  }
]
```

**Reply by ssyberg** [reply]
Ok I actually figured out whats happening here, it looks like it's only traversing that first `row` fieldset, it doesn't appear to be looking at anything else.

**Reply by manuelebagnolini** [reply]
Hi @ssyberg,  thank you for giving us this example, we have released version 0.4 which should solve the problem you have encountered.

We have also added a first implementation for displaying the percentage of document translation as discussed in #2577 

In the language bar of the document a colored indicator is shown which indicates the completion status of the texts in the language and a tooltip appears with the explicit percentage by hovering over the language button:

![image](https://github.com/payloadcms/payload/assets/45173102/4aaa7325-dd92-448e-808d-51f9fab4afef)

A known issue is that at the moment the calculation does not evaluate texts inside `blocks` fields but that will be fixed in the next version.

**Reply by ssyberg** [reply]
Awesome, so far so good, will report back if we hit any issues. JFYI looks like the css import is still required, might want to add that back to the config instructions for the moment (https://www.npmjs.com/package/@websolutespa/payload-plugin-localization)

**Reply by manuelebagnolini** [reply]
Thank you! We have updated the instructions specifying the correct import of the css.

**Comment by manuelebagnolini** [comment]
> Hi @manuelebagnolini , thank you for your plugin! It's very helpful! I would like to try adding a button that allows me to translate the text using ChatGPT, but I can't open your code repository. Will you share the repo?

Hi @simonereda we will definitely share the repo, at the moment it is still private because the plugin is part of a mono repo where the rest of the project is not yet possible to make it public.

Furthermore this plugin is still an alpha version that still needs more testing.

**Comment by ssyberg** [comment]
Hi @manuelebagnolini the more I've been using this the more I've come to feel the mix of localized and non-localized fields on a given edit view is going to be very confusing for our editors. As such, I'd like to suggest something that might seem a bit drastic - what if on all "translation" views, non localized fields were fully disabled and only enabled on the "default" language view? I think the UX would need to be thought through a bit, but IMO would be a big improvement overall.
